### ISSUE TITLE
Changing an EV/NV should also change other EV/NV

### ISSUE THREAD

**Issue by SvenRosvall** [david284/ModuleDescriptor#40]
This was reported by Greg Palmer when testing MMC with CANSERVO8C.

CANSERVO8C (and CANPAN) have a "type" bit in an EV that indicates if an event is produced or consumed. Changing this bit changes how other EV's are interpreted by the firmware. This may lead to inconsistencies and confusing behaviour. Some EVs may have values that cannot be represented by the Module Descriptor File for that "event type". This may happen for an option field where the value does not have an option or the value is outside its min-max range. 

Suggested fixes include making this "type" bit read-only once the event is created, or add a rule feature that is triggered when the "type" bit changes and when triggered sets some EV's to sensible values for that event "type".

**Comment by SvenRosvall** [comment]
I think this is a rare edge case. Once an event is created you won't normally change the event "type". 

Suggest marking this as low priority or even "won't fix" but keep it open to record that this inconsistency can happen.

**Comment by RichardCrawshaw** [comment]
I think this is an MMC issue not an MDF issue. 

Config tools **may** be helped by additional data in the MDF such as defining default values under certain conditions, but how to handle changes is a config tool decision.

**Comment by SvenRosvall** [comment]
Richard. Yes, I was thinking about logging this issue against MMC. But this is a issue for any configuration tool and I believe these config tools need support from the descriptor file anyway. So this is why I logged it here instead of MMC **and** your tool. 

But as I noted above, I believe this is low priority or even a thing that shouldn't be fixed.

**Comment by RichardCrawshaw** [comment]
I think we've talked about this elsewhere: was it private emails? I can't rightly remember. Anyway...

There are a number of things that a config tool can do when a Variable changes:
1. Do nothing; let any invalid or out of range values remain: it is up to the user to alter them.
1. Read all Variables from the Module.
1. Read a sub-set of the Variables from the Module (affected Variables).
1. Set certain Variables within the local data to specific values (to align with values that the Module itself has set those Variables to). 

The items 3 & 4 need module specific data to know which Variables / values. This is where MDF data comes in, though it could also help with items 1 & 2 also.

Item 1 is the default. Most Variables don't affect other Variables to this extent. No additional data is really needed, but for completeness there could be an explicit value that indicates this.

Item 2 is fairly straightforward. It would just need a Variable level setting, possibly down to the bit level, that indicates that all other Variables should be read from the Module after a change. I'm less happy with this approach as it's a big hammer solution. Reading all Variables takes a noticeable amount of time and creates a large amount of bus traffic, but under certain conditions it may be the best solution.

For Item 3 there are three MDF approaches could be taken:
- Against the Affecting Variable, for each value that will initiate a read, list the Affected Variables: those that need to be read.
- Against each of the Affected Variables list the Affecting Variables and the corresponding value(s): those that cause them to need to be read.
- Have a separate section that lists both sets of Variables together. If there are multiple groups of affecting and affected Variables then have multiple sets of lists. This is probably the most compact set of data, and probably also the easiest to interpret, but it could be considered to be a departure from how the MDF is currently structured.

Item 4 is similar to Item 3. It just has the addition of a value against each Affected Variable, but that does increase the complexity of the data. Re-listing the approaches:
- Against the Affecting Variable, for each value that will initiate a value change, list tuples of the Affected Variable and the new value .
- Against each of the Affected Variables list tuples of the Affecting Variable, its value and the new value to assign.
- A separate section that lists the affecting and affected Variables and the new value for each. This is more complex as there may be a different value for each affected Variable for each affecting Variable.

I'm happy to work up a more detailed data structure if wanted.
